src/drv_oculus_rift/rift.c

```diff
@@ -139,24 +222,213 @@ static void handle_tracker_sensor_msg(rift_priv* priv, unsigned char* buffer, in
 		vec3f_from_rift_vec(s->samples[i].accel, &priv->raw_accel);
 		vec3f_from_rift_vec(s->samples[i].gyro, &priv->raw_gyro);
 
+#if 0
+		printf ("HMD dt %f a [%f,%f,%f], g[%f,%f,%f]\n",
+		    dt, priv->raw_accel.x, priv->raw_accel.y, priv->raw_accel.z,
+				priv->raw_gyro.x, priv->raw_gyro.y, priv->raw_gyro.z);
+#endif
 		ofusion_update(&priv->sensor_fusion, dt, &priv->raw_gyro, &priv->raw_accel, &priv->raw_mag);
 		dt = TICK_LEN; // TODO: query the Rift for the sample rate
 	}
 
 	priv->last_imu_timestamp = s->timestamp;
 }
 
-static void update_device(ohmd_device* device)
+static void
+handle_touch_controller_message(rift_hmd_t *hmd,
+		rift_touch_controller_t *touch, pkt_rift_radio_message *msg)
+{
+	// The top bits are carrying something unknown. Ignore them
+  uint8_t buttons = msg->touch.buttons & 0xf;
+	rift_touch_calibration *c = &touch->calibration;
+
+	if (touch->buttons != buttons) {
+		LOGV ("touch controller %d buttons now %x",
+				touch->base.id, buttons);
+	}
+	touch->buttons = buttons;
+
+	if (!(msg->touch.timestamp ||
+	      msg->touch.accel[0] || msg->touch.accel[1] || msg->touch.accel[2] ||
+	      msg->touch.gyro[0] || msg->touch.gyro[1] || msg->touch.gyro[2]))
+		return;
+
+	if (!touch->have_calibration) {
+		/* We need calibration data to do any more */
+		if (rift_touch_get_calibration (hmd->radio_handle, touch->device_num,
+				&touch->calibration) < 0)
+			return;
+		touch->have_calibration = true;
+	}
+
+	// time in microseconds
+	int32_t dt;
+	if (touch->time_valid)
+		dt = msg->touch.timestamp - touch->last_timestamp;
+	else
+		dt = 0;
+
+	const double dt_s = 1e-6 * dt;
+	double a[3] = {
+		OHMD_GRAVITY_EARTH / 2048 * msg->touch.accel[0],
+		OHMD_GRAVITY_EARTH / 2048 * msg->touch.accel[1],
+		OHMD_GRAVITY_EARTH / 2048 * msg->touch.accel[2],
+	};
+	double g[3] = {
+		2.0 / 2048 * msg->touch.gyro[0],
+		2.0 / 2048 * msg->touch.gyro[1],
+		2.0 / 2048 * msg->touch.gyro[2],
+	};
+	vec3f mag = {{0.0f, 0.0f, 0.0f}};
+	vec3f gyro;
+	vec3f accel;
+
+	/* Apply correction offsets first - bottom row of the
```

Comment:
Just a note, this doesn't seem to fit well for my controllers. Given that none of the other obvious variations seem to work either, let's go with what works for yours.

Reply:
I'll get some data from your controllers from you